File: spring_java/src/main/java/dev/pollito/spring_java/sakila/auth/adapter/in/rest/AuthRestMapper.java
package dev.pollito.spring_java.sakila.auth.adapter.in.rest;

import dev.pollito.spring_java.config.security.userdetails.SakilaUserDetails;
import java.util.List;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.springframework.security.core.GrantedAuthority;

@Mapper(componentModel = "spring")
public interface AuthRestMapper {

  @Mapping(
      target = "authorities",
      expression = "java(mapAuthorities(sakilaUserDetails.getAuthorities()))")
  dev.pollito.spring_java.sakila.generated.model.UserDetails map(
      SakilaUserDetails sakilaUserDetails);

  default List<String> mapAuthorities(
      java.util.Collection<? extends GrantedAuthority> authorities) {
    if (authorities == null) {
      return null;
    }
    return authorities.stream().map(GrantedAuthority::getAuthority).toList();
  }
}

File: spring_kotlin/src/main/kotlin/dev/pollito/spring_kotlin/sakila/auth/adapter/in/rest/AuthRestMapper.kt
package dev.pollito.spring_kotlin.sakila.auth.adapter.`in`.rest

import dev.pollito.spring_kotlin.config.security.userdetails.SakilaUserDetails
import dev.pollito.spring_kotlin.sakila.generated.model.UserDetails
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.Named
import org.springframework.security.core.GrantedAuthority

@Mapper(componentModel = "spring")
interface AuthRestMapper {

  @Mapping(source = "username", target = "username")
  @Mapping(source = "accountNonLocked", target = "accountNonLocked")
  @Mapping(source = "authorities", target = "authorities", qualifiedByName = ["mapAuthorities"])
  @Mapping(source = "staff", target = "staff")
  fun map(sakilaUserDetails: SakilaUserDetails): UserDetails

  @Named("mapAuthorities")
  fun mapAuthorities(authorities: Collection<GrantedAuthority>): List<String> {
    return authorities.mapNotNull { it.authority }
  }
}
